### PR TITLE
new face_detection package with dlib library

### PR DESCRIPTION
**Pull request by Lizca**

### CODE REVIEW

File: dlib_face_detection/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+## Install ##
+#############
+install(TARGETS dlib_face_detection head_detector viewer face_detection_merge
+	ARCHIVE DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
+	LIBRARY DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
+	RUNTIME DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION}
+)
+
+install(DIRECTORY include
+	DESTINATION ${CATKIN_GLOBAL_INCLUDE_DESTINATION}
```
Comment (fmessmer) [inline dlib_face_detection/CMakeLists.txt]: `CATKIN_PACKAGE_INCLUDE_DESTINATION`

File: dlib_face_detection/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+target_link_libraries(face_detection_merge ${catkin_LIBRARIES})
+ 
+set_target_properties(head_detector PROPERTIES COMPILE_FLAGS -D__LINUX__)
+add_dependencies(head_detector ${catkin_EXPORTED_TARGETS} ${${PROJECT_NAME}_EXPORTED_TARGETS})
+
+
+#############
+## Install ##
+#############
+install(TARGETS dlib_face_detection head_detector viewer face_detection_merge
```
Comment (fmessmer) [inline dlib_face_detection/CMakeLists.txt]: add `data_file`

File: dlib_face_detection/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+add_executable(viewer src/detection_viewer.cpp include)
+add_executable(face_detection_merge src/face_detection_merge.cpp include)
+
+target_link_libraries(dlib_face_detection ${catkin_LIBRARIES} ${OpenCV_LIBS} ${X11_LIBRARIES} dlib data_file)
+target_link_libraries(head_detector ${catkin_LIBRARIES} ${OpenCV_LIBS})
+target_link_libraries(viewer ${catkin_LIBRARIES} ${OpenCV_LIBS} ${X11_LIBRARIES})
+target_link_libraries(face_detection_merge ${catkin_LIBRARIES})
+ 
+set_target_properties(head_detector PROPERTIES COMPILE_FLAGS -D__LINUX__)
+add_dependencies(head_detector ${catkin_EXPORTED_TARGETS} ${${PROJECT_NAME}_EXPORTED_TARGETS})
```
Comment (fmessmer) [inline dlib_face_detection/CMakeLists.txt]: add_dependencies(head_detector ${${PROJECT_NAME}_EXPORTED_TARGETS} ${catkin_EXPORTED_TARGETS})

and add to all the other `add_executable`, too

File: dlib/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+## Install ##
+#############
+install(TARGETS dlib
+	ARCHIVE DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
+	LIBRARY DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
+	RUNTIME DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION}
+)
+
+install(DIRECTORY common/include
+	DESTINATION ${CATKIN_GLOBAL_INCLUDE_DESTINATION}
```
Comment (fmessmer) [inline dlib/CMakeLists.txt]: `CATKIN_PACKAGE_INCLUDE_DESTINATION`

File: dlib/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+)
+install(DIRECTORY /usr/include
+	DESTINATION ${CATKIN_GLOBAL_INCLUDE_DESTINATION}
+)
```
Comment (fmessmer) [inline dlib/CMakeLists.txt:51]: this installs all the headers from `/usr/include`...you proably don't want that...and I think you actually do not need it

File: dlib/common/include/text.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+
```
Comment (fmessmer) [inline dlib/common/include/text.txt]: empy file?

File: dlib/package.xml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+<package format="2">
+	<name>dlib_lib</name>
+	<version>0.1.0</version>
+	<description>
+		This package wrappes the external c++ library dlib (http://dlib.net/) in a ros package, so other packages can use it. It downloads the source code of it and then builds it, corresponding to the webpage of it, using cmake. The build library is a static library and will be copied to the created common/lib folder. The neccessary header files will be copied to the common/include folder. The include folder needs to exist befor building the package, because in the CMakeLists you have to specify where the headers are and if the folder doesn't exist, catkin won't be able to build the package. For further descriptions and tutorials see the Makefile.tarball and http://dlib.net/ .
```
Comment (fmessmer) [inline dlib/package.xml]: wraps

File: dlib_face_detection/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+find_package(catkin REQUIRED COMPONENTS
+  cob_perception_msgs
+  cob_vision_utils
+  cv_bridge
+  dlib_lib
+  message_generation
+  pcl_ros
+  roscpp
+  std_msgs
+  tf
```
Comment (fmessmer) [inline dlib_face_detection/CMakeLists.txt:16]: these are not all dependencies, e.g. `message_filter` is missing

also adjust the `package.xml` accordingly

File: dlib_face_detection/include/dlib_face_detection/detection_viewer.h [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+#include <message_filters/subscriber.h>
+#include <message_filters/synchronizer.h>
+#include <message_filters/sync_policies/approximate_time.h>
+
+// opencv
+#include <opencv/cv.h>
+#include <opencv/ml.h>
+#include <cv_bridge/cv_bridge.h>
+#include <opencv2/imgproc/imgproc.hpp>
+#include <opencv/highgui.h>
```
Comment (fmessmer) [inline dlib_face_detection/include/dlib_face_detection/detection_viewer.h:26]: `opencv` side-by-side with `opencv2`? does this not conflict?

File: dlib_face_detection/package.xml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  
+  <depend>cob_perception_msgs</depend>
+  <depend>cob_vision_utils</depend>
+  <depend>cv_bridge</depend>
+  <depend>dlib_lib</depend>
+  <depend>pcl_ros</depend>
+  <depend>roscpp</depend>
+  <depend>sensor_msgs</depend>
+  <depend>std_msgs</depend> 
+  <depend>tf</depend>
```
Comment (fmessmer) [inline dlib_face_detection/package.xml]: finalize dependencies

File: dlib_face_detection/src/detection_viewer.cpp [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+
+DetectionViewer::DetectionViewer(ros::NodeHandle nh)
+{
+	image_sub_ = new message_filters::Subscriber<image_msgT>(nh, "image_in", 10);
+	head_sub_ = new message_filters::Subscriber<det_msgT>(nh, "/people_detections/head_detections", 10);
+	face_sub_ = new message_filters::Subscriber<det_msgT>(nh, "/people_detections/face_detections", 10);
+
+	sync_ = new message_filters::Synchronizer<SyncPolicy>(SyncPolicy(30), *image_sub_, *head_sub_, *face_sub_);
+	sync_->registerCallback(boost::bind(&DetectionViewer::callback, this, _1, _2, _3));
```
Comment (ipa-fxm) [inline dlib_face_detection/src/detection_viewer.cpp]: indentation

File: dlib_face_detection/src/detection_viewer.cpp [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+    cv::namedWindow("dlib Detections", cv::WINDOW_AUTOSIZE );
+    cv::startWindowThread();
```
Comment (fmessmer) [inline dlib_face_detection/src/detection_viewer.cpp]: indentation

File: dlib_face_detection/src/dlib_face_detection.cpp [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+				rect.x = head_img_msg->head_detections[j].head_detection.x;
+				rect.y = head_img_msg->head_detections[j].head_detection.y;
+				rect.width = head_img_msg->head_detections[j].head_detection.width;
+				rect.height = head_img_msg->head_detections[j].head_detection.height;
+
+				rectArray->rects.push_back(rect);
+			}
+		}
+
+		head_based_face_pub_.publish(rectArray);
```
Comment (fmessmer) [inline dlib_face_detection/src/dlib_face_detection.cpp]: indentation

File: dlib_face_detection/src/face_detection_merge.cpp [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+				break;
+			}
+
+			 // publish tf
+			 transform_.setOrigin(tf::Vector3(pose_world.pose.position.x, pose_world.pose.position.y, pose_world.pose.position.z));
+			 tf::Quaternion q;
+			 q.setRPY(pose_world.pose.orientation.x, pose_world.pose.orientation.y, pose_world.pose.orientation.z);
+			 transform_.setRotation(q);
+
+			 tfb_.sendTransform(tf::StampedTransform(transform_, pose_world.header.stamp, pose_world.header.frame_id, s.str()));
```
Comment (fmessmer) [inline dlib_face_detection/src/face_detection_merge.cpp]: indentation

**Comment by floweisshardt** [comment]
@Lizca: Please fix @ipa-fxm's comments and check if it is building on travis

File: dlib_face_detection/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+  ${BZIP_INCLUDE_DIRS}
+  include
+  
+)
+
+add_library(data_file ${EXTRACTED_FILE})
+set_target_properties(data_file PROPERTIES LINKER_LANGUAGE CXX)
+target_link_libraries(data_file ${BZIP2_LIBRARIES})
+
+add_executable(dlib_face_detection src/dlib_face_detection.cpp include)
```
Comment (fmessmer) [inline dlib_face_detection/CMakeLists.txt]: remove the `include` from the `add_executable` tags...they are not needed

File: dlib/common/include/text.txt (hunk 6 above)
Comment (Lizca) [inline dlib/common/include/text.txt]: This file is important because git will delete the "common/include" folder otherwise and they have to exist before build.

**Comment by Lizca** [comment]
It seems like the build problems are solved...

File: dlib/common/include/text.txt (hunk 6 above)
Comment (fmessmer) [inline dlib/common/include/text.txt]: with a little google'ing....it seems to be a git "convention" to the `touch .gitkeep` within that folder

File: dlib/package.xml [hunk 16]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+<package format="2">
+	<name>dlib_lib</name>
+	<version>0.1.0</version>
+	<description>
+		This package wrapes the external c++ library dlib (http://dlib.net/) in a ros package, so other packages can use it. It downloads the source code of it and then builds it, corresponding to the webpage of it, using cmake. The build library is a static library and will be copied to the created common/lib folder. The neccessary header files will be copied to the common/include folder. The include folder needs to exist befor building the package, because in the CMakeLists you have to specify where the headers are and if the folder doesn't exist, catkin won't be able to build the package. For further descriptions and tutorials see the Makefile.tarball and http://dlib.net/ .
```
Comment (fmessmer) [inline dlib/package.xml]: wraps

File: dlib_face_detection/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,110 @@
+cmake_minimum_required(VERSION 2.8.3)
+project(dlib_face_detection)
+
+find_package(catkin REQUIRED COMPONENTS
+  cob_people_detection
+  cob_perception_msgs
+  cob_vision_utils
+  cv_bridge
+  dlib_lib
```
Comment (fmessmer) [inline dlib_face_detection/CMakeLists.txt]: isn't the package just call `dlib`?

File: dlib_face_detection/CMakeLists.txt [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+  OUTPUT "${EXTRACTED_FILE}"
+  PRE_LINK
+  COMMAND bunzip2 -dkf ${PROJECT_SOURCE_DIR}/data/shape_predictor_68_face_landmarks.dat.bz2
+  WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}/data
+  COMMENT "Unpacking ${DOWNLOAD_PATH}"
+)
+
+catkin_package(
+  INCLUDE_DIRS
+    ../cob_people_detection/common/include
```
Comment (fmessmer) [inline dlib_face_detection/CMakeLists.txt]: you cannot have a relative path to another package's include files this in here!
As this package actually only contains executables, you probably can just put `catkin_package()`, i.e. nothing inside the brackets

File: dlib_face_detection/CMakeLists.txt [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+add_library(data_file ${EXTRACTED_FILE})
+set_target_properties(data_file PROPERTIES LINKER_LANGUAGE CXX)
+target_link_libraries(data_file ${BZIP2_LIBRARIES})
+
+
+add_executable(dlib_face_detection src/dlib_face_detection.cpp)
+target_link_libraries(dlib_face_detection ${catkin_LIBRARIES} ${OpenCV_LIBS} ${X11_LIBRARIES} dlib data_file)
+add_dependencies(dlib_face_detection ${${PROJECT_NAME}_EXPORTED_TARGETS} ${catkin_EXPORTED_TARGETS})
+
+add_executable(head_detector src/head_detector_node.cpp ../cob_people_detection/common/src/head_detector.cpp)
```
Comment (fmessmer) [inline dlib_face_detection/CMakeLists.txt]: no relative path to other packages cpp file!
why do you need to compile against `cob_people_detection` code file?

File: dlib_face_detection/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+target_link_libraries(head_detector ${catkin_LIBRARIES} ${OpenCV_LIBS})
+add_dependencies(head_detector ${${PROJECT_NAME}_EXPORTED_TARGETS} ${catkin_EXPORTED_TARGETS})
+set_target_properties(head_detector PROPERTIES COMPILE_FLAGS -D__LINUX__)
+
+add_executable(viewer src/detection_viewer.cpp include)
```
Comment (fmessmer) [inline dlib_face_detection/CMakeLists.txt]: remove include

File: dlib_face_detection/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+target_link_libraries(viewer ${catkin_LIBRARIES} ${OpenCV_LIBS} ${X11_LIBRARIES})
+add_dependencies(viewer ${${PROJECT_NAME}_EXPORTED_TARGETS} ${catkin_EXPORTED_TARGETS})
+
+add_executable(face_detection_merge src/face_detection_merge.cpp include)
```
Comment (fmessmer) [inline dlib_face_detection/CMakeLists.txt]: remove include

File: dlib_face_detection/src/detection_viewer.cpp [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+	cv::startWindowThread();
+}
+
+DetectionViewer::~DetectionViewer()
+{
+}
+
+void DetectionViewer::callback(const sensor_msgs::Image::ConstPtr &image_msg,
+								const cob_perception_msgs::DetectionArray::ConstPtr& head_msg,
+								const cob_perception_msgs::DetectionArray::ConstPtr& face_msg)
```
Comment (fmessmer) [inline dlib_face_detection/src/detection_viewer.cpp]: indentation (could be reduced)

File: dlib_face_detection/src/dlib_face_detection.cpp [hunk 23]
```diff
@@ -0,0 +1,114 @@
+
+#include <dlib_face_detection/dlib_face_detection.h>
+
+
+FaceDetectorNode::FaceDetectorNode(ros::NodeHandle nh)
+{
+	dlib::deserialize(ros::package::getPath("dlib_face_detection") + "/data/shape_predictor_68_face_landmarks.dat") >> pose_model_;
```
Comment (fmessmer) [inline dlib_face_detection/src/dlib_face_detection.cpp:8]: this file is not part of the PR

File: dlib_face_detection/src/dlib_face_detection.cpp [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+	node_handle_.param("scale_factor", scale_factor_, 2.0);
+
+	detector_ = dlib::get_frontal_face_detector();
+
+	head_img_sub_ = nh.subscribe<cob_perception_msgs::ColorDepthImageArray>("/dlib_face_detection/head_detections", 2,  &FaceDetectorNode::headCallback, this);
+	image_sub_ = nh.subscribe<sensor_msgs::Image>("image_in", 2, &FaceDetectorNode::imgCallback, this);
+
+	direct_face_pub_ = node_handle_.advertise<dlib_face_detection::RectArray>("/dlib_face_detection/direct_face_detections", 1);
+
+	head_based_face_pub_ = node_handle_.advertise<dlib_face_detection::RectArray>("/dlib_face_detection/head_based_face_detections", 1);
```
Comment (fmessmer) [inline dlib_face_detection/src/dlib_face_detection.cpp:19]: avoid global namespaces for publisher and subscriber!

File: dlib_face_detection/src/face_detection_merge.cpp [hunk 25]
```diff
@@ -0,0 +1,190 @@
+
+#include <dlib_face_detection/face_detection_merge.h>
+
+FaceDetectionMerge::FaceDetectionMerge(ros::NodeHandle nh)
+{
+	node_handle_.param("enable_tf", enable_tf_, false);
+
+	cloud_sub_ = new message_filters::Subscriber<cloud_msgT>(nh, "pointcloud_in", 10);
+	direct_face_sub_ = new message_filters::Subscriber<rect_msgT>(nh, "/dlib_face_detection/direct_face_detections", 10);
+	head_based_face_sub_ = new message_filters::Subscriber<rect_msgT>(nh, "/dlib_face_detection/head_based_face_detections", 10);
```
Comment (fmessmer) [inline dlib_face_detection/src/face_detection_merge.cpp:12]: avoid global namespaces for pubisher and subscriber

File: dlib_face_detection/src/face_detection_merge.cpp [hunk 26]
```diff
@@ -0,0 +1,190 @@
...
+
+		int a = 0;
+		while (a < 5)
+		{
+			// transform pose to "/odom_combined" (world coordinates)
+			try
+			{
+				ros::Time now = ros::Time(0);
+				tfl_.waitForTransform("/odom_combined", detArray->header.frame_id, now, ros::Duration(0.5));
+				tfl_.transformPose("/odom_combined", now, detArray->detections[i].pose, "/odom_combined", pose_world);
```
Comment (fmessmer) [inline dlib_face_detection/src/face_detection_merge.cpp]: do we really want to transform against `odom_combined`? maybe `base_link` is better as it is "on" the robot and not affected by navigation/localization...

**Comment by fmessmer** [comment]
some of my comments apply to several code files, e.g. "global namespaces" or "odom_combined"...did not want to put it everywhere...please have a look wherever it applies...

File: dlib_face_detection/src/dlib_face_detection.cpp (hunk 23 above)
Comment (Lizca) [inline dlib_face_detection/src/dlib_face_detection.cpp:8]: it will be downloaded during build

File: dlib_face_detection/CMakeLists.txt (hunk 17 above)
Comment (Lizca) [inline dlib_face_detection/CMakeLists.txt]: oh i see ... the folder is called `dlib` but the package in it `dlib_lib`. 
I will change the package name into `dlib` too

File: dlib_face_detection/src/face_detection_merge.cpp (hunk 26 above)
Comment (floweisshardt) [inline dlib_face_detection/src/face_detection_merge.cpp]: we do not need to transfrom at all. please publish tf in camera frame

File: dlib_face_detection/include/dlib_face_detection/detection_viewer.h (hunk 9 above)
Comment (Lizca) [inline dlib_face_detection/include/dlib_face_detection/detection_viewer.h:26]: I haven't detected any problems with that and the Travis build was clean too. So i think there are no conflicts.

**Comment by NadiaHG** [comment]
@ipa-rmb could you please review this PR?

**Comment by ipa-rmb** [comment]
Do we use dlib? Do you know whether there is any face detection or recognition function delivered by dlib which is superior to ours? If you do not know, I will check that. If there is no additional benefit or need for dlib, I would not like to add it to the repository as maintenance would be more intricate given this external lib. If we do not merge it to the offical main branch, we can still "save" the dlib integration work in a separate branch.

**Comment by floweisshardt** [comment]
we do not use dlib at the moment. I'm not aware of any mayor differences. Saving this in a separate branch is ok for me.
@ipa-rmb: it's your decision

**Comment by ipa-rmb** [comment]
Yes, then let's have a separate branch for this. How can I change the target branch of the pull request?

**Comment by floweisshardt** [comment]
you can't. We'll have to close this PR, clone the repository from Lizca, push to a branch on your fork and then open a new PR. 

Let me know if you need help

**Comment by ipa-rmb** [comment]
redirected into a separate archive branch in PR #50 .
